idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,774 @@
+
+         Returns:
+             None
+        '''
+
+        self._update_UT_DT(self.model,implemented_shut_down, implemented_start_up)
+        self._update_power(self.model,implemented_power_output)
+
+        return
+
+    # define a function to add bidding constraints to models
+    @staticmethod
+    def _add_bidding_constraints(m,NA_con_range = 24):
+
+        '''
+        This function takes the thermal model and add bidding cosntraints
+        to it in order to bid into DAM.
+
+        Arguments:
+            m: the bidding thermal generator model
+            NA_con_range: number of time steps this constraint will be enforced.
+
+        Returns:
+            None
+        '''
+
+        # generate scenarios combinations
+        scenario_comb = combinations(m.SCENARIOS,2)
+
+        # constraints for thermal generators
+        def bidding_con_fun1(m):
+            for k in scenario_comb:
+                for j in m.UNITS:
+                    for h in range(NA_con_range):
+                        yield (m.P_T[j,h,k[0]] - m.P_T[j,h,k[1]])\
+                        *(m.DAM_price[h,k[0]] - m.DAM_price[h,k[1]]) >= 0
+        m.bidding_con1 = pyo.ConstraintList(rule = bidding_con_fun1)
+
+        return
+
+    # define a function to add non_anticipativity constraints to models
+    @staticmethod
+    def add_non_anticipativity(m,NA_con_range = 2):
+
+        '''
+        This function takes the thermal model and add non non_anticipativity
+        cosntraints to it in order to do self-scheduling in DAM.
+
+        Arguments:
+            m: the bidding thermal generator model
+            NA_con_range: number of time steps this constraint will be enforced.
+
+        Returns:
+            None
+        '''
+
+        # generate scenarios combinations
+        scenario_comb = combinations(m.SCENARIOS,2)
+
+        def NAC_list_rules(m):
+            for k in scenario_comb:
+                for j in m.UNITS:
+                    for h in range(NA_con_range):
+                        yield m.gen_pow[j,h,k[0]] == m.gen_pow[j,h,k[1]]
+        m.NAC_SS_cons = ConstraintList(rule = NAC_list_rules)
+
+        return m
+
+    def stochastic_bidding(self,date):
+
+        '''
+        This methods solve the stochastic bidding optimization problem given the
+        date we are bidding into.
+
+        Arguments:
+            date: the date we are bidding into
+
+        Returns:
+            bids: the bid we computed. It is a dictionary that has this structure
+            {t: {gen:{power: cost}}}.
+        '''
+
+        print("")
+        print("In stochastic_bidding\n")
+
+        m = self.model
+
+        # select the price forecasts
+        forecasts = self.price_forecasts.loc[date].values
+
+        # pass the forecast into pyomo model
+        for i in m.SCENARIOS:
+            for t in m.HOUR:
+                m.DAM_price[t,i] = forecasts[t,i]
+
+        # solve the model
+        solver = pyo.SolverFactory('gurobi')
```

Comment:
Make solver live in self. That way the user can define a custom solver. If `solver=None` in `__init__`, then use Gurobi as the default. We'll need to use CBC for automated testing.